File: infra/lib/database-stack.ts
import * as cdk from "aws-cdk-lib";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import { Construct } from "constructs";
import { config, expensesTableName } from "./config";

interface DatabaseStackProps extends cdk.StackProps {
  environment: string;
}

/**
 * DynamoDB table for expenses, with a string `id` partition key. The table name
 * is composed as `<environment>-expenses`; the backend derives the same name
 * from the ENVIRONMENT env var (BackendStack).
 */
export class DatabaseStack extends cdk.Stack {
  public readonly table: dynamodb.Table;

  constructor(scope: Construct, id: string, props: DatabaseStackProps) {
    super(scope, id, props);

    this.table = new dynamodb.Table(this, "ExpensesTable", {
      tableName: expensesTableName(props.environment),
      partitionKey: {
        name: config.partitionKey,
        type: dynamodb.AttributeType.STRING,
      },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
  }
}

File: infra/stacks/database_stack.py
import aws_cdk as cdk
from aws_cdk import aws_dynamodb as dynamodb
from constructs import Construct

from .config import Config, expense_status_table_name, expenses_table_name


class DatabaseStack(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        environment: str,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.table = dynamodb.Table(
            self,
            "ExpensesTable",
            table_name=expenses_table_name(environment),
            partition_key=dynamodb.Attribute(
                name=Config.PARTITION_KEY,
                type=dynamodb.AttributeType.STRING,
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            removal_policy=cdk.RemovalPolicy.DESTROY,
        )

        # Per-month paid/due status for recurring expense instances (issue #10):
        # keyed by a composite "{expense_id}#{month}" id so a recurring
        # expense's paid status is tracked independently for every month it
        # recurs into.
        self.status_table = dynamodb.Table(
            self,
            "ExpenseStatusTable",
            table_name=expense_status_table_name(environment),
            partition_key=dynamodb.Attribute(
                name=Config.PARTITION_KEY,
                type=dynamodb.AttributeType.STRING,
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            removal_policy=cdk.RemovalPolicy.DESTROY,
        )
